src/api/java/com/minecolonies/api/util/InventoryUtils.java

```diff
@@ -1482,4 +1494,107 @@ public static boolean removeStacksFromProvider(final ICapabilityProvider provide
 
         return true;
     }
+
+    /**
+     * Check if a certain item is in the provider but without the provider being full.
+     * @param provider the provider to check.
+     * @param item the item.
+     * @param itemDamage its damage.
+     * @return the slot or -1.
+     */
+    public static int findFirstSlotInProviderWithNotFull(final ICapabilityProvider provider, final Item item, final int itemDamage)
+    {
+        for (IItemHandler handler : getItemHandlersFromProvider(provider))
+        {
+            int foundSlot = findFirstSlotInItemHandlerWithNotFull(handler, (ItemStack stack) -> compareItems(stack, item, itemDamage));
+            //TODO: When contract is hardened later: Replace this -1 check with a try-catch block.
+            if (foundSlot > -1)
+            {
+                return foundSlot;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Check if a certain item is in the handler but without the provider being full.
+     * @param handler the handler to check.
+     * @param itemStackSelectionPredicate the selection predicate..
+     * @return the slot or -1.
+     */
+    public static int findFirstSlotInItemHandlerWithNotFull(final IItemHandler handler, @NotNull final Predicate<ItemStack> itemStackSelectionPredicate)
+    {
+        boolean foundEmptySlot = false;
+        boolean foundItem = false;
+        int itemSlot = -1;
+        for (int slot = 0; slot < handler.getSlots(); slot++)
+        {
+            final ItemStack stack = handler.getStackInSlot(slot);
+            if(ItemStackUtils.isEmpty(stack))
+            {
+                foundEmptySlot = true;
+            }
+            else if (itemStackSelectionPredicate.test(stack))
+            {
+                foundItem = true;
+                itemSlot = slot;
+            }
+
+            if(foundItem == foundEmptySlot)
+            {
+                return itemSlot;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Spawn an itemStack in the world.
+     * @param worldIn the world.
+     * @param x the x pos.
+     * @param y the y pos.
+     * @param z the z pos.
+     * @param stack the stack to drop.
+     */
+    public static void spawnItemStack(final World worldIn, final double x, final double y, final double z, final ItemStack stack)
+    {
+        final Random random = new Random();
+        final double f = random.nextDouble() * SPAWN_MODIFIER + SPAWN_ADDITION;
```

Comment:
This can be named properly.
f, f1, f2, are the In-Block Position of the EntityItem. As such the be named properly.